File: C/Day-18/problem-1/problem-1-solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

#define GRID_SIZE 71
#define MAX_INPUT 3000
#define INF INT_MAX

// Direction vectors for moving up, down, left, or right
int dx[] = {-1, 1, 0, 0};
int dy[] = {0, 0, -1, 1};

// Function to check if a position is within bounds and not corrupted
int is_valid(int x, int y, int grid[GRID_SIZE][GRID_SIZE], int visited[GRID_SIZE][GRID_SIZE]) {
    return x >= 0 && x < GRID_SIZE && y >= 0 && y < GRID_SIZE && grid[x][y] == 0 && !visited[x][y];
}

// BFS to find the shortest path from (0, 0) to (70, 70)
int bfs(int grid[GRID_SIZE][GRID_SIZE]) {
    int visited[GRID_SIZE][GRID_SIZE] = {0};
    int queue[MAX_INPUT][3]; // queue to store x, y, and distance
    int front = 0, rear = 0;

    // Start BFS from (0, 0)
    queue[rear][0] = 0;
    queue[rear][1] = 0;
    queue[rear][2] = 0; // Distance
    rear++;
    visited[0][0] = 1;

    while (front < rear) {
        int x = queue[front][0];
        int y = queue[front][1];
        int dist = queue[front][2];
        front++;

        // Check if we reached the destination
        if (x == GRID_SIZE - 1 && y == GRID_SIZE - 1) {
            return dist;
        }

        // Explore all 4 directions
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (is_valid(nx, ny, grid, visited)) {
                queue[rear][0] = nx;
                queue[rear][1] = ny;
                queue[rear][2] = dist + 1;
                rear++;
                visited[nx][ny] = 1;
            }
        }
    }

    return -1; // If no path is found
}

int main() {
    FILE *file = fopen("inputdata.txt", "r");
    if (!file) {
        perror("Error opening input file");
        return EXIT_FAILURE;
    }

    int grid[GRID_SIZE][GRID_SIZE] = {0};
    int x, y;

    // Read up to 1024 bytes from the input file and mark them as corrupted
    for (int i = 0; i < 1024 && fscanf(file, "%d,%d", &x, &y) == 2; i++) {
        if (x >= 0 && x < GRID_SIZE && y >= 0 && y < GRID_SIZE) {
            grid[x][y] = 1; // Mark the position as corrupted
        }
    }

    fclose(file);

    // Find the shortest path using BFS
    int result = bfs(grid);

    if (result != -1) {
        printf("The minimum number of steps needed to reach the exit is: %d\n", result);
    } else {
        printf("It is not possible to reach the exit.\n");
    }

    return 0;
}

File: Go/18/problem-1/problem-1-solution.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const GridSize = 71 // 0-70 for both X and Y
const MaxBytes = 1024

// Direction vectors for moving up, down, left, or right
var directions = [][2]int{
	{-1, 0}, // Up
	{1, 0},  // Down
	{0, -1}, // Left
	{0, 1},  // Right
}

func readInput(filename string) ([][2]int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var bytes [][2]int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		coords := strings.Split(line, ",")
		if len(coords) != 2 {
			continue
		}
		x, err1 := strconv.Atoi(coords[0])
		y, err2 := strconv.Atoi(coords[1])
		if err1 != nil || err2 != nil {
			continue
		}
		bytes = append(bytes, [2]int{x, y})
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return bytes, nil
}

func bfs(grid [][]bool) int {
	// Queue for BFS: each entry contains (x, y, steps)
	type state struct {
		x, y, steps int
	}
	queue := []state{{0, 0, 0}}

	// Visited array to avoid revisiting
	visited := make([][]bool, GridSize)
	for i := range visited {
		visited[i] = make([]bool, GridSize)
	}
	visited[0][0] = true

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		// If we reach the bottom-right corner, return the number of steps
		if current.x == GridSize-1 && current.y == GridSize-1 {
			return current.steps
		}

		// Explore neighbors
		for _, dir := range directions {
			nx, ny := current.x+dir[0], current.y+dir[1]

			// Check bounds and whether the cell is corrupted or visited
			if nx >= 0 && ny >= 0 && nx < GridSize && ny < GridSize && !grid[nx][ny] && !visited[nx][ny] {
				visited[nx][ny] = true
				queue = append(queue, state{nx, ny, current.steps + 1})
			}
		}
	}

	// If no path found, return -1
	return -1
}

func main() {
	// Read the input data
	bytes, err := readInput("inputdata.txt")
	if err != nil {
		fmt.Println("Error reading input file:", err)
		return
	}

	// Initialize the grid
	grid := make([][]bool, GridSize)
	for i := range grid {
		grid[i] = make([]bool, GridSize)
	}

	// Mark corrupted cells for the first MaxBytes entries
	for i := 0; i < MaxBytes && i < len(bytes); i++ {
		x, y := bytes[i][0], bytes[i][1]
		if x >= 0 && x < GridSize && y >= 0 && y < GridSize {
			grid[x][y] = true
		}
	}

	// Find the shortest path using BFS
	steps := bfs(grid)

	if steps == -1 {
		fmt.Println("No path to the exit.")
	} else {
		fmt.Printf("The minimum number of steps to reach the exit is: %d\n", steps)
	}
}
